Cannot build solidity with gcc-15.0.1 -O3 but builds just fine with -O2

## Description

I'm trying to build solidity with GCC 15.0.1 with O3 (default) and with 02 (default in Fedora Linux) on x86_64. The latter builds just fine, but the former halts with the following error

## Environment

- Compiler version: gcc 15.0.1
- Compilation pipeline (legacy, IR, EOF): n/a (compile-time issue)
- Target EVM version (as per compiler settings): n/a (compile-time issue)
- Framework/IDE (e.g. Foundry, Hardhat, Remix): n/a (compile-time issue)
- EVM execution environment / backend / blockchain client: n/a (compile-time issue)
- Operating system: Fedora Linux 43

## Steps to Reproduce

```
CFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -mbranch-protection=standard -fasynchronous-unwind-tables -fstack-clash-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer '
CXXFLAGS='-O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -mbranch-protection=standard -fasynchronous-unwind-tables -fstack-clash-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer '
LDFLAGS='-Wl,-z,relro -Wl,--as-needed  -Wl,-z,pack-relative-relocs -Wl,-z,now -specs=/usr/lib/rpm/redhat/redhat-hardened-ld -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -Wl,--build-id=sha1 -specs=/usr/lib/rpm/redhat/redhat-package-notes '
LT_SYS_LIBRARY_PATH=/usr/lib64:
CC=gcc
CXX=g++
/usr/bin/cmake -S . -B redhat-linux-build -DCMAKE_C_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_CXX_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_Fortran_FLAGS_RELEASE:STRING=-DNDEBUG -DCMAKE_VERBOSE_MAKEFILE:BOOL=ON -DCMAKE_INSTALL_DO_STRIP:BOOL=OFF -DCMAKE_INSTALL_PREFIX:PATH=/usr -DCMAKE_INSTALL_FULL_SBINDIR:PATH=/usr/bin -DCMAKE_INSTALL_SBINDIR:PATH=bin -DCMAKE_POLICY_VERSION_MINIMUM=3.5 -DBUILD_SHARED_LIBS:BOOL=ON -DIGNORE_VENDORED_DEPENDENCIES:BOOL=ON -DBoost_USE_STATIC_LIBS:BOOL=OFF -DSTRICT_Z3_VERSION:BOOL=OFF -DTESTS:BOOL=OFF
...
[ 20%] Building CXX object libevmasm/CMakeFiles/evmasm.dir/SimplificationRules.cpp.o
cd /builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/redhat-linux-build/libevmasm && /usr/bin/g++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DFMT_HEADER_ONLY=1 -I/builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/redhat-linux-build/include -I/builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29 -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-U_FORTIFY_SOURCE,-D_FORTIFY_SOURCE=3 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -mbranch-protection=standard -fasynchronous-unwind-tables -fstack-clash-protection -fno-omit-frame-pointer -mno-omit-leaf-frame-pointer -O3 -DNDEBUG -std=c++20 -fstack-protector-strong -Wimplicit-fallthrough -fmacro-prefix-map=/builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29=/solidity -Wpessimizing-move -Wredundant-move -Wall -Wextra -Werror -pedantic -Wmissing-declarations -Wno-unknown-pragmas -Wsign-conversion -Wconversion -Wextra-semi -Wno-dangling-reference -Wduplicated-cond -Wlogical-op -fdiagnostics-color -MD -MT libevmasm/CMakeFiles/evmasm.dir/SimplificationRules.cpp.o -MF CMakeFiles/evmasm.dir/SimplificationRules.cpp.o.d -o CMakeFiles/evmasm.dir/SimplificationRules.cpp.o -c /builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/libevmasm/SimplificationRules.cpp
In file included from /builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/libevmasm/Assembly.h:23,
                 from /builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/libevmasm/Inliner.h:25,
                 from /builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/libevmasm/Inliner.cpp:23:
In member function ‘solidity::evmasm::AssemblyItem::AssemblyItem(solidity::evmasm::AssemblyItem const&)’,
    inlined from ‘std::_Construct<solidity::evmasm::AssemblyItem, solidity::evmasm::AssemblyItem const&>(solidity::evmasm::AssemblyItem*, solidity::evmasm::AssemblyItem const&)void’ at /usr/include/c++/15/bits/stl_construct.h:133:7,
    inlined from ‘std::__do_uninit_copy<solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem*>(solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem*)solidity::evmasm::AssemblyItem*’ at /usr/include/c++/15/bits/stl_uninitialized.h:145:17,
    inlined from ‘std::uninitialized_copy<solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem*>(solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem*)solidity::evmasm::AssemblyItem*’ at /usr/include/c++/15/bits/stl_uninitialized.h:260:30,
    inlined from ‘std::__uninitialized_copy_a<solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem*, solidity::evmasm::AssemblyItem>(solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem*, std::allocator<solidity::evmasm::AssemblyItem>&)solidity::evmasm::AssemblyItem*’ at /usr/include/c++/15/bits/stl_uninitialized.h:617:32,
    inlined from ‘std::vector<solidity::evmasm::AssemblyItem, std::allocator<solidity::evmasm::AssemblyItem> >::_M_range_initialize<solidity::evmasm::AssemblyItem const*>(solidity::evmasm::AssemblyItem const*, solidity::evmasm::AssemblyItem const*, std::forward_iterator_tag)void’ at /usr/include/c++/15/bits/stl_vector.h:1890:6,
    inlined from ‘std::vector<solidity::evmasm::AssemblyItem, std::allocator<solidity::evmasm::AssemblyItem> >::vector(std::initializer_list<solidity::evmasm::AssemblyItem>, std::allocator<solidity::evmasm::AssemblyItem> const&)’ at /usr/include/c++/15/bits/stl_vector.h:709:21,
    inlined from ‘solidity::evmasm::Inliner::shouldInlineFullFunctionBody(unsigned long, ranges::span<solidity::evmasm::AssemblyItem const, -1l>, unsigned long) const’ at /builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/libevmasm/Inliner.cpp:159:2:
/builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/libevmasm/AssemblyItem.h:163:9: error: ‘MEM[(const struct AssemblyItem &)&D.578956].m_instruction’ may be used uninitialized [-Werror=maybe-uninitialized]
  163 |         AssemblyItem(AssemblyItem const&) = default;
      |         ^~~~~~~~~~~~
/builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/libevmasm/Inliner.cpp: In member function ‘solidity::evmasm::Inliner::shouldInlineFullFunctionBody(unsigned long, ranges::span<solidity::evmasm::AssemblyItem const, -1l>, unsigned long) const’:
/builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/libevmasm/Inliner.cpp:159:9: note: ‘<anonymous>’ declared here
  159 |         };
      |         ^
cc1plus: all warnings being treated as errors
gmake[2]: *** [libevmasm/CMakeFiles/evmasm.dir/build.make:236: libevmasm/CMakeFiles/evmasm.dir/Inliner.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[2]: Leaving directory '/builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/redhat-linux-build'
gmake[1]: Leaving directory '/builddir/build/BUILD/solidity-0.8.29-build/solidity-0.8.29/redhat-linux-build'
gmake[1]: *** [CMakeFiles/Makefile2:453: libevmasm/CMakeFiles/evmasm.dir/all] Error 2
gmake: *** [Makefile:139: all] Error 2
```

--- a/libevmasm/AssemblyItem.h
+++ b/libevmasm/AssemblyItem.h
@@ -105,7 +105,6 @@ public:
 
 	explicit AssemblyItem(bytes _verbatimData, size_t _arguments, size_t _returnVariables):
 		m_type(VerbatimBytecode),
-		m_instruction{},
 		m_verbatimBytecode{{_arguments, _returnVariables, std::move(_verbatimData)}},
 		m_debugData{langutil::DebugData::create()}
 	{}
@@ -195,7 +194,7 @@ public:
 	/// @returns true if the item has m_instruction properly set.
 	bool hasInstruction() const
 	{
-		return
+		bool const shouldHaveInstruction =
 			m_type == Operation ||
 			m_type == EOFCreate ||
 			m_type == ReturnContract ||
@@ -206,12 +205,14 @@ public:
 			m_type == RetF ||
 			m_type == SwapN ||
 			m_type == DupN;
+		solAssert(shouldHaveInstruction == m_instruction.has_value());
+		return shouldHaveInstruction;
 	}
 	/// @returns the instruction of this item (only valid if hasInstruction returns true)
 	Instruction instruction() const
 	{
 		solAssert(hasInstruction());
-		return m_instruction;
+		return *m_instruction;
 	}
 
 	/// @returns true if the type and data of the items are equal.
@@ -323,7 +324,7 @@ private:
 	size_t opcodeCount() const noexcept;
 
 	AssemblyItemType m_type;
-	Instruction m_instruction; ///< Only valid if m_type == Operation
+	std::optional<Instruction> m_instruction; ///< Only valid for item types that represent a specific opcode
 	std::shared_ptr<u256> m_data; ///< Only valid if m_type != Operation
 	std::optional<FunctionSignature> m_functionSignature; ///< Only valid if m_type == CallF or JumpF
 	/// If m_type == VerbatimBytecode, this holds number of arguments, number of

